### PR TITLE
Add gptneox conversion script

### PR DESCRIPTION
**Pull request by ZanezZephyrs**
Adding the gptneox conversion script, the script has a "original_implementation" flag that allows us to convert a t5x checkpoint to the original gptneox architecture ( useful for some experiments). 

Currently this is a draft because i still need to change the script to not copy the bias in the last layernorm, waiting for the first checkpoint of exp44 to test

### CODE REVIEW

File: t5x/scripts/convert_gptneox_t5x_to_hf.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+from t5x import checkpoints
+
+
+def patch_config_file(hf_checkpoint_path):
+    config_path = Path(hf_checkpoint_path, "config.json")
+    config_data = json.load(open(config_path))
+
+    config_data["layer_norm_eps"] = 1e-06
+
+    out_file = open(config_path, "w")
```
Comment (rodrigonogueira4) [inline t5x/scripts/convert_gptneox_t5x_to_hf.py:24]: use `with open`?

File: t5x/scripts/convert_gptneox_t5x_to_hf.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    json.dump(config_data, out_file, indent=4)
+
+
+def reconstruct_qkv_kernel(query, key, value, config):
```
Comment (rodrigonogueira4) [inline t5x/scripts/convert_gptneox_t5x_to_hf.py:28]: add description explaining why we need this function?

File: t5x/scripts/convert_gptneox_t5x_to_hf.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+            torch.tensor(t5x_mlp_wo_weight.get().transpose(), dtype=torch.float16)
+        )
+        gptneox_model.gpt_neox.layers[layer_index].mlp.dense_4h_to_h.bias = torch.nn.Parameter(
+            torch.tensor(t5x_mlp_wo_bias.get(), dtype=torch.float16)
+        )
+
+    # Decoder Normalization
+    tx5_decoder_norm_weight = t5x_model["target"]["decoder"]["decoder_norm"]["scale"]
+    # TODO: move this bias to the original condition
+    tx5_decoder_norm_bias = t5x_model["target"]["decoder"]["decoder_norm"]["bias"]
```
Comment (rodrigonogueira4) [inline t5x/scripts/convert_gptneox_t5x_to_hf.py:153]: We don't have this for our final version, right? (e.g., exp44)

File: t5x/scripts/convert_gptneox_t5x_to_hf.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        torch.tensor(tx5_decoder_norm_bias.get(), dtype=torch.float16)
+    )
+
+    # Token Embeddings
+    tx5_lm_head_weight = t5x_model["target"]["decoder"]["token_embedder"]["embedding"]
+    gptneox_model.gpt_neox.embed_in.weight = torch.nn.Parameter(
+        torch.tensor(tx5_lm_head_weight.get(), dtype=torch.float16)
+    )
+
+    if "logits_dense" in t5x_model["target"]["decoder"]:
```
Comment (rodrigonogueira4) [inline t5x/scripts/convert_gptneox_t5x_to_hf.py:168]: remove as we _should_ have this layer?

File: t5x/scripts/convert_gptneox_t5x_to_hf.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+from transformers import GPTNeoXConfig, GPTNeoXForCausalLM
+
+from t5x import checkpoints
+
+
+def patch_config_file(hf_checkpoint_path):
+    config_path = Path(hf_checkpoint_path, "config.json")
+    config_data = json.load(open(config_path))
+
+    config_data["layer_norm_eps"] = 1e-06
```
Comment (rodrigonogueira4) [inline t5x/scripts/convert_gptneox_t5x_to_hf.py:22]: Also add
```
config_data["tokenizer_class"] = "GPT2Tokenizer"
```
like here?
https://huggingface.co/EleutherAI/gpt-j-6B/blob/main/config.json#L36

File: t5x/scripts/download_and_evaluate_checkpoints.py [hunk 6]
```diff
@@ -253,7 +253,11 @@ def get_not_processed_checkpoints(available_checkpoints, results_path):
                 f"CONVERSION SKIP - DIR {converted_checkpoint_save_path} already exists, skipping checkpoint conversion, you might wanna to delete the folder to convert the checkpoint again"
             )
         else:
-            base_command_for_t5x_checkpoint_conversion = f"cd {PATH_FOR_T5X_TO_HF} && python3 convert_gptj_t5x_to_hf.py"
+
+            if config.get("model_type",None) == "gptneox":
+                base_command_for_t5x_checkpoint_conversion = f"cd {PATH_FOR_T5X_TO_HF} && python3 convert_gptneox_t5x_to_hf.py"
+            else:
```
Comment (rodrigonogueira4) [inline t5x/scripts/download_and_evaluate_checkpoints.py:259]: nit: add 
```
elif "gpt-j":
    ...
else:
    raise Exception(f'model_type "{model_type}" not implemented')
```

File: t5x/scripts/convert_gptneox_t5x_to_hf.py (hunk 3 above)
Comment (ZanezZephyrs) [reply t5x/scripts/convert_gptneox_t5x_to_hf.py:153]: yeah, this was there because i was still converting the exp 43, i am going to move it to the original condition

File: t5x/scripts/convert_gptneox_t5x_to_hf.py (hunk 2 above)
Comment (ZanezZephyrs) [reply t5x/scripts/convert_gptneox_t5x_to_hf.py:29]: done

File: t5x/scripts/convert_gptneox_t5x_to_hf.py (hunk 5 above)
Comment (ZanezZephyrs) [reply t5x/scripts/convert_gptneox_t5x_to_hf.py:22]: added, i already did this in my last experiment and as expected, now the autotokenizer works properly

File: t5x/scripts/download_and_evaluate_checkpoints.py (hunk 6 above)
Comment (ZanezZephyrs) [reply t5x/scripts/download_and_evaluate_checkpoints.py:265]: added

File: t5x/scripts/calculate_mc4_loss.py [hunk 7]
```diff
@@ -49,14 +49,25 @@
         help="number of sequences to use for evaluation",
     )
 
+    parser.add_argument(
+        "--fp32",
+        action="store_true",
```
Comment (rodrigonogueira4) [inline t5x/scripts/calculate_mc4_loss.py:54]: I thought that we would be using fp16 by default?

File: t5x/scripts/convert_gptneox_t5x_to_hf.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+    config_data["tokenizer_class"] = "GPT2Tokenizer"
+
+    out_file = open(config_path, "w")
+    json.dump(config_data, out_file, indent=4)
+
+
+def reconstruct_qkv_kernel(query, key, value, config):
+    """builds the qkv weigth matrix used in the gptneox HF implementation
+    from the query, value and key arrays used in the t5x implementation
```
Comment (rodrigonogueira4) [inline t5x/scripts/convert_gptneox_t5x_to_hf.py:31]: I think the most important part here is to say that we merge the query, value, key tensors into a single tensor of shape [3*dim, dim].

**Review by rodrigonogueira4 — APPROVED** [review]

**Comment by rodrigonogueira4** [comment]
Thanks a lot for doing this!

File: t5x/scripts/calculate_mc4_loss.py (hunk 7 above)
Comment (ZanezZephyrs) [reply t5x/scripts/calculate_mc4_loss.py:54]: we are, we only use fp32 if the "--fp32" flag is provided.

Comment (rodrigonogueira4) [reply t5x/scripts/calculate_mc4_loss.py:54]: Ah sorry!

File: t5x/scripts/convert_gptneox_t5x_to_hf.py [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+        )
+        gptneox_model.gpt_neox.layers[layer_index].mlp.dense_4h_to_h.weight = torch.nn.Parameter(
+            torch.tensor(t5x_mlp_wo_weight.get().transpose(), dtype=torch.float16)
+        )
+        gptneox_model.gpt_neox.layers[layer_index].mlp.dense_4h_to_h.bias = torch.nn.Parameter(
+            torch.tensor(t5x_mlp_wo_bias.get(), dtype=torch.float16)
+        )
+
+    # Decoder Normalization
+    tx5_decoder_norm_weight = t5x_model["target"]["decoder"]["decoder_norm"]["scale"]
```
Comment (hugoabonizio) [inline t5x/scripts/convert_gptneox_t5x_to_hf.py:173]:
```suggestion
    t5x_decoder_norm_weight = t5x_model["target"]["decoder"]["decoder_norm"]["scale"]
```

File: t5x/scripts/convert_gptneox_t5x_to_hf.py [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+
+    gptneox_model.gpt_neox.final_layer_norm.weight = torch.nn.Parameter(
+        torch.tensor(tx5_decoder_norm_weight.get(), dtype=torch.float16)
+    )
+
+    if original_implementation:
+        # the original implementation has bias in the last layer norm
+        tx5_decoder_norm_bias = t5x_model["target"]["decoder"]["decoder_norm"]["bias"]
```
Comment (hugoabonizio) [inline t5x/scripts/convert_gptneox_t5x_to_hf.py:181]:
```suggestion
        t5x_decoder_norm_bias = t5x_model["target"]["decoder"]["decoder_norm"]["bias"]
```

File: t5x/scripts/convert_gptneox_t5x_to_hf.py [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+
+        gptneox_model.gpt_neox.final_layer_norm.bias = torch.nn.Parameter(
+            torch.tensor(tx5_decoder_norm_bias.get(), dtype=torch.float16)
```
Comment (hugoabonizio) [inline t5x/scripts/convert_gptneox_t5x_to_hf.py:184]:
```suggestion
            torch.tensor(t5x_decoder_norm_bias.get(), dtype=torch.float16)
```

File: t5x/scripts/convert_gptneox_t5x_to_hf.py [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+        )
+    # Token Embeddings
+    tx5_lm_head_weight = t5x_model["target"]["decoder"]["token_embedder"]["embedding"]
```
Comment (hugoabonizio) [inline t5x/scripts/convert_gptneox_t5x_to_hf.py:187]:
```suggestion
    t5x_lm_head_weight = t5x_model["target"]["decoder"]["token_embedder"]["embedding"]
```

File: t5x/scripts/convert_gptneox_t5x_to_hf.py [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+    gptneox_model.gpt_neox.embed_in.weight = torch.nn.Parameter(
+        torch.tensor(tx5_lm_head_weight.get(), dtype=torch.float16)
```
Comment (hugoabonizio) [inline t5x/scripts/convert_gptneox_t5x_to_hf.py:189]:
```suggestion
        torch.tensor(t5x_lm_head_weight.get(), dtype=torch.float16)
```

File: t5x/scripts/convert_gptneox_t5x_to_hf.py [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+    )
+
+    tx5_token_embeddings = t5x_model["target"]["decoder"]["logits_dense"]["kernel"]
```
Comment (hugoabonizio) [inline t5x/scripts/convert_gptneox_t5x_to_hf.py:192]:
```suggestion
    t5x_token_embeddings = t5x_model["target"]["decoder"]["logits_dense"]["kernel"]
```

File: t5x/scripts/convert_gptneox_t5x_to_hf.py [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+    gptneox_model.embed_out.weight = torch.nn.Parameter(
+        torch.tensor(tx5_token_embeddings.get().transpose(), dtype=torch.float16)
```
Comment (hugoabonizio) [inline t5x/scripts/convert_gptneox_t5x_to_hf.py:194]:
```suggestion
        torch.tensor(t5x_token_embeddings.get().transpose(), dtype=torch.float16)
```

File: t5x/scripts/convert_gptneox_t5x_to_hf.py [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+        )
+        gptneox_model.gpt_neox.layers[layer_index].mlp.dense_4h_to_h.bias = torch.nn.Parameter(
+            torch.tensor(t5x_mlp_wo_bias.get(), dtype=torch.float16)
+        )
+
+    # Decoder Normalization
+    tx5_decoder_norm_weight = t5x_model["target"]["decoder"]["decoder_norm"]["scale"]
+
+    gptneox_model.gpt_neox.final_layer_norm.weight = torch.nn.Parameter(
+        torch.tensor(tx5_decoder_norm_weight.get(), dtype=torch.float16)
```
Comment (hugoabonizio) [inline t5x/scripts/convert_gptneox_t5x_to_hf.py:176]:
```suggestion
        torch.tensor(t5x_decoder_norm_weight.get(), dtype=torch.float16)
```

File: t5x/scripts/convert_gptneox_t5x_to_hf.py [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+    gptneox_model.save_pretrained(hf_checkpoint_path)
+    # adjust some values in the model config
+    patch_config_file(hf_checkpoint_path)
+    print("GPTNEOX Model was sucessfully converted!")
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    # Required parameters
+    parser.add_argument("--t5x_checkpoint_path", default=None, type=str, required=True, help="Path the TX5 checkpoint.")
```
Comment (hugoabonizio) [inline t5x/scripts/convert_gptneox_t5x_to_hf.py:206]:
```suggestion
    parser.add_argument("--t5x_checkpoint_path", default=None, type=str, required=True, help="Path the T5X checkpoint.")
```
